examples/permissioned-token/assets/controller.py

```diff
@@ -0,0 +1,254 @@
+    Expects 3 arguments:
+    * add_permission : name of the branch
+    * permission_id : permissions smart contract application index
+    * permissions_manager : premissions manager address
+    NOTE: token_id is also passed in txn.ForeignAssets (to load it's asset manager)
+    """
+    add_new_permission = Seq([
+        assetManager, # load asset_manager (from Store) of Txn.ForeignAssets[0]
+        Assert(And(
+    		Txn.application_args.length() == Int(3),
+            # Txn.assets.length() == Int(1), [TEALv3]
+    		# Txn.assets[0] == App.globalGet(token_id), [TEALv3]
+
+    		Txn.sender() == assetManager.value(),
+        )),
+
+        # Update total rules counter (= 1 in this case) as we only have a single
+        # permissions contract
+        App.globalPut(var_total, App.globalGet(var_total) + Int(1)),
+
+        # Add permissions(rules) smart contract config in global state (permission_id, permissions_manager)
+        App.globalPut(permission_id, Btoi(Txn.application_args[1])),
+        App.globalPut(permissions_manager, Txn.application_args[2]),
+        Return(Int(1))
+    ])
+
+    """
+    Change permissions manager of permissions ssc. Only accepted if sender is asset manager.
+    Expects 1 argument in Txn.accounts[n] array:
+    * addr : address of the new permissions manager
+    """
+    change_permissions_manager = Seq([
+        assetManager,
+        Assert(And(
+    		Txn.application_args.length() == Int(1),
+    		# Txn.assets[0] == App.globalGet(token_id), [TEALv3]
+
+    		Txn.sender() == assetManager.value(),
+        )),
+
+        # update permissions manager address to the first account address passed in appAccounts
+        App.globalPut(permissions_manager, Txn.accounts[1]),
+        Return(Int(1))
+    ])
+
+    # Check basic tx (call to controller, clawback tx)
+    verify_basic_calls = And(
+        # verify first transaction
+        # call to controller smart contract - signed by asset sender
+        Txn.group_index() == Int(0), # this tx (call to controller) should be 1st in group
+        Gtxn[0].application_id() == Global.current_application_id(),
+
+        # verify 2nd tx
+        Gtxn[1].type_enum() == TxnType.AssetTransfer, # this should be clawback call (to transfer asset)
+        Gtxn[1].xfer_asset() == App.globalGet(token_id), # check asset_id passed through params
+    )
+
+    # Check permissions smart contract is called and it's application index is correct
+    verify_perm_call = And(
+        # verify rules call (ensure permissions smart contract is being called in 4th tx)
+        Gtxn[3].type_enum() == TxnType.ApplicationCall,
+        Gtxn[3].application_id() == App.globalGet(permission_id),
+    )
+
+    # verifies that asset sender in 2nd tx
+    # - calls the controller smart contract
+    # - is also the sender of the payment tx (to pay fees of clawback escrow)
+    # - calls the permissions smart contract (ensures rules check)
```

Comment:
let's move the comments to the statements below:
```
# asset sender is also the payment tx sender (to pay fees of clawback escrow)
Gtxn[0].sender() == Gtxn[2].sender(),
...
```

Reply:
alright